lightning/src/ln/onion_utils.rs

```diff
@@ -935,6 +935,160 @@ pub(crate) fn decode_next_payment_hop<NS: Deref>(
 	}
 }
 
+/// Build a payment onion, returning the first hop msat and cltv values as well.
+pub fn create_payment_onion<T>(
+	secp_ctx: &Secp256k1<T>, path: &Path, session_priv: &SecretKey, total_msat: u64, 
+	recipient_onion: RecipientOnionFields, starting_htlc_offset: u32, payment_hash: PaymentHash,
+	keysend_preimage: Option<PaymentPreimage>, prng_seed: [u8; 32]) -> Result<(u64, u32, msgs::OnionPacket), ()>
+where
+	T: secp256k1::Signing
+{
+	let onion_keys = construct_onion_keys(&secp_ctx, &path, &session_priv).map_err(|_| ())?;
+	let (onion_payloads, htlc_msat, htlc_cltv) = build_onion_payloads(
+        &path,
+        total_msat,
+        recipient_onion,
+        starting_htlc_offset,
+        &keysend_preimage,
+    ).map_err(|_| ())?;
+	let onion_packet = construct_onion_packet(onion_payloads, onion_keys, prng_seed, &payment_hash)?;
+	Ok((htlc_msat, htlc_cltv, onion_packet))
+}
+
+/// Forwarded Payment, including next hop details
+#[derive(Debug)]
+pub struct ForwardedPayment {
+	/// short channel id of the next hop
+	pub short_channel_id: u64,
+	/// The value, in msat, of the payment after this hop's fee is deducted.
+	pub amt_to_forward: u64,
+	/// outgoing CLTV for the next hop
+	pub outgoing_cltv_value: u32,
+	/// onion packet for the next hop
+	pub onion_packet: msgs::OnionPacket,
+}
+
+/// Received payment, of either regular or blinded type
+#[derive(Debug)]
+pub enum ReceivedPayment {
+	/// Regular (unblinded) payment
+	Regular {
+		/// payment_secret to authenticate sender to the receiver
+		payment_secret: Option<[u8; 32]>,
+		/// The total value, in msat, of the payment as received by the ultimate recipient
+		total_msat: Option<u64>,
+		/// custom payment metadata included in the payment
```

Comment:
Throughout this commit, please capitalize the start of sentences and end them with a period:
```suggestion
		/// Custom payment metadata included in the payment.
```